File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    // public int lengthOfLIS(int[] nums) {
    //     int n = nums.length;
    //     int dp[] = new int[n];
    //     int ans = 0;
    //     for(int i=0;i<n;i++){
    //         dp[i] = 1;
    //         for(int j=0;j<i;j++){
    //             if(nums[j] < nums[i]){
    //                 dp[i] = Math.max(dp[i],dp[j] + 1);
    //             }
    //         }
    //         ans = Math.max(ans,dp[i]);
    //     }
    //     return ans;
    // }
    public int lengthOfLIS(int[] nums) {
        int[] tails = new int[nums.length];
        int size = 0;
        for (int x : nums) {
            int i = 0, j = size;
            while (i != j) {
                int m = (i + j) / 2;
                if (tails[m] < x)
                    i = m + 1;
                else
                    j = m;
            }
            tails[i] = x;
            if (i == size) ++size;
        }
        return size;
    }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        if(n == 0) return 0;
        
        vector<int> lis(n,1);
        int maxlis = 1;
        for(int i=1;i<n;i++) {
            int j = i - 1;
            while(j>=0) {
                if(nums[j] < nums[i]) {
                    lis[i] = max(lis[i],lis[j] + 1);
                }
                j--;
            }
            
            maxlis = max(maxlis,lis[i]);
        }
        int ans = maxlis;
        for(int i=n-1;i>=0;i--) {
            if(lis[i] == maxlis) {
                cout << nums[i] << " ";
                maxlis--;
            }
        } 
        cout << "\n";
        return ans;
        
    }
};